src/qibo/quantum_info/quantum_networks.py

```diff
@@ -629,53 +565,497 @@ def _run_checks(self, partition, system_output, pure):
+            calculations. If ``None``, defaults to :class:`qibo.backends.GlobalBackend`.
+            Defaults to ``None``.
+    """
+
+    def __init__(
+        self,
+        tensor,
+        partition: Optional[Union[List[int], Tuple[int]]] = None,
+        *,
+        system_input: Optional[Union[List[bool], Tuple[bool]]] = None,
+        pure: bool = False,
+        backend=None,
+    ):
+        if partition == None:
+            if pure:
+                partition = tensor.shape
+            else:
+                partition = tuple(int(np.sqrt(d)) for d in tensor.shape)
+        if len(partition) % 2 != 0:
+            raise_error(
+                ValueError,
+                "A quantum comb should only contain equal number of input and output systems. "
+                + "For general quantum networks, one should use the ``QuantumNetwork`` class.",
             )
+        if system_input is not None:
+            warning("system_input is ignored for QuantumComb")
 
-        if self.system_output is None:
-            self.system_output = [
-                True,
-            ] * len(self.partition)
-            for k in range(len(self.partition) // 2):
-                self.system_output[k * 2] = False
-            self.system_output = tuple(self.system_output)
+        super().__init__(
+            tensor, partition, [True, False] * (len(partition) // 2), pure, backend
+        )
+
+    def is_causal(
+        self, order: Optional[Union[int, str]] = None, precision_tol: float = 1e-8
+    ):
+        """Returns bool indicating if the Choi operator :math:`\\mathcal{J}` of the network satisfies the causal order condition.
+
+        Causality is calculated based on a recursive constrains.
+        This method reduce a n-comb to a (n-1)-comb at each step, and checks if the reduced comb is independent on the
+        last output system.
+
+        Args:
+            order (str or int, optional): order of the norm. Defaults to ``None``.
+            precision_tol (float, optional): threshold :math:`\\epsilon` that defines
+                if Choi operator of the network is :math:`\\epsilon`-close to causality
+                in the norm given by ``order``. Defaults to :math:`10^{-8}`.
+
+        Returns:
+            bool: Causal order condition.
+        """
+        if precision_tol < 0.0:
+            raise_error(
+                ValueError,
+                f"``precision_tol`` must be non-negative float, but it is {precision_tol}",
+            )
+
+        if order is None and self._backend.__class__.__name__ == "TensorflowBackend":
+            order = "euclidean"
+
+        backend = self._backend
+
+        dim_out = self.partition[-1]
+        dim_in = self.partition[-2]
+
+        trace_out = backend.cast(
+            trace(dim_out, backend=backend).full(), dtype=self._tensor.dtype
+        )
+        trace_in = backend.cast(
+            trace(dim_in, backend=backend).full(), dtype=self._tensor.dtype
+        )
```

Comment:
If `trace` accepts `backend` as argument, then the resulting object should already be returned in the correct backend and the `cast` outside should not be needed.